mount/titus-mount-lustre.c

```diff
@@ -0,0 +1,359 @@
+	if (ret == -1) {
+		perror("setns user");
+		return 1;
+	}
+	ret = setns(mnt_fd, CLONE_NEWNS);
+	if (ret == -1) {
+		perror("setns mnt");
+		return 1;
+	}
+	ret = setns(net_fd, CLONE_NEWNET);
+	if (ret == -1) {
+		perror("setns net");
+		return 1;
+	}
+	fsfd = fsopen("lustre", 0);
+	if (fsfd == -1) {
+		perror("fsopen");
+		return 1;
+	}
+	assert(send_fd(sk, fsfd) == 0);
+	return 0;
+}
+
+static int fork_and_get_fsfd(long int nsfd)
+{
+	int sk_pair[2], ret, fsfd, status;
+	pid_t worker;
+
+	if (socketpair(PF_LOCAL, SOCK_SEQPACKET, 0, sk_pair) < 0) {
+		perror("socketpair");
+		exit(1);
+	}
+	worker = fork();
+	if (worker < 0) {
+		perror("fork");
+		exit(1);
+	}
+	if (worker == 0) {
+		close(sk_pair[0]);
+		ret = setup_fsfd_in_namespaces(sk_pair[1], nsfd);
+		close(sk_pair[1]);
+		exit(ret);
+	}
+	close(sk_pair[1]);
+	fsfd = recv_fd(sk_pair[0]);
+	assert(fsfd >= 0);
+	if (waitpid(worker, &status, 0) != worker) {
+		perror("waitpid");
+		exit(1);
+	}
+	if (!WIFEXITED(status) || WEXITSTATUS(status)) {
+		fprintf(stderr, "worker exited nonzero\n");
+		exit(1);
+	}
+	close(sk_pair[0]);
+	return fsfd;
+}
+
+static void switch_namespaces(int nsfd)
+{
+	int ret;
+	int mnt_fd = openat(nsfd, "mnt", O_RDONLY);
+	assert(mnt_fd != -1);
+	ret = setns(mnt_fd, CLONE_NEWNS);
+	if (ret == -1) {
+		perror("setns mount");
+		exit(1);
+	}
+	int net_fd = openat(nsfd, "net", O_RDONLY);
+	assert(net_fd != -1);
+	ret = setns(net_fd, CLONE_NEWNET);
+	if (ret == -1) {
+		perror("setns net");
+		exit(1);
+	}
+}
+
+static void mount_and_move(int fsfd, const char *target,
+			   unsigned long flags)
+{
+	int mfd = fsmount(fsfd, 0, flags);
+	if (mfd < 0) {
+		mount_error(fsfd, "fsmount");
+		E(close(fsfd));
+	}
+	E(move_mount(mfd, "", AT_FDCWD, target, MOVE_MOUNT_F_EMPTY_PATH));
+	E(close(mfd));
+}
+
+int main(int argc, char *argv[])
+{
+	int nsfd, fsfd, res = -1;
+	unsigned long flags_ul;
+	/*
+	 * We do this because parsing args is a bigger pain than passing
+	 * via environment variable, although passing via environment
+	 * variable has a "cost" in that they are limited in size
+	 */
+	const char *pidDir = getenv("TITUS_PID1_DIR");
+	const char *lustre_mount_hostname = getenv("MOUNT_LUSTRE_HOSTNAME");
+    const char *lustre_mount_mountname = getenv("MOUNT_LUSTRE_MOUNTNAME");
+	const char *target = getenv("MOUNT_TARGET");
+	const char *flags = getenv("MOUNT_FLAGS");
+	const char *options = getenv("MOUNT_OPTIONS");
+
+	int buf_size = sysconf(_SC_PAGESIZE);
+	char final_options[buf_size];
+	strcpy(final_options, options);
+
+	if (!(target && flags && options && pidDir)) {
+		fprintf(stderr,
+			"Usage: must provide MOUNT_TARGET, MOUNT_FLAGS, TITUS_PID1_DIR, and MOUNT_OPTIONS env vars");
+		return 1;
+	}
+
+	if (!(lustre_mount_hostname && lustre_mount_mountname)) {
+		fprintf(stderr,
+			"Usage: must provide MOUNT_LUSTRE_HOSTNAME, MOUNT_LUSTRE_MOUNTNAME env vars");
+		return 1;
+	}
+
+	errno = 0;
+	flags_ul = strtoul(flags, NULL, 10);
+	if (errno) {
```

Comment:
The man page says "The  implementation  may also set errno to EINVAL in case no conversion was performed (no digits seen, and 0 returned)." so maybe we need a check for flags_ul != 0 here too?

Reply:
That is weird. `0` is very normal (no extra flags).
But if literally *no* digits were seen, then yea that would be an err and should be reported already on that line

Reply:
Yeah, I guess it's really strlen(flags) == 0 too; the implementation is free to *not* report an error in the zero-length case, or it might. But actually 0 seems like a fine default for no mount flags specified, so I don't think there's any change necessary here.